opensearch-operator/pkg/reconcilers/ismpolicy.go

```diff
@@ -0,0 +1,521 @@
+		reason = "waiting for opensearch cluster status to be running"
+		r.recorder.Event(r.instance, "Normal", opensearchPending, reason)
+		retResult = ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 10 * time.Second,
+		}
+		return
+	}
+
+	r.osClient, err = util.CreateClientForCluster(r.ctx, r.Client, r.cluster, r.osClientTransport)
+	if err != nil {
+		reason := "error creating opensearch client"
+		r.recorder.Event(r.instance, "Warning", opensearchError, reason)
+	}
+
+	// Check ism policy state to make sure we don't touch preexisting ism policy
+	if r.instance.Status.ExistingISMPolicy == nil {
+		var exists bool
+		exists, retErr = services.PolicyExists(r.ctx, r.osClient, r.instance.Spec.PolicyID)
+		if retErr != nil {
+			reason = "failed to get policy status from Opensearch API"
+			r.logger.Error(retErr, reason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			return
+		}
+		if pointer.BoolDeref(r.updateStatus, true) {
+			retErr = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+					return err
+				}
+				r.instance.Status.ExistingISMPolicy = &exists
+				return r.Status().Update(r.ctx, r.instance)
+			})
+			if retErr != nil {
+				reason = fmt.Sprintf("failed to update status: %s", retErr)
+				r.recorder.Event(r.instance, "Warning", statusError, reason)
+				return
+			}
+		} else {
+			// Emit an event for unit testing assertion
+			r.recorder.Event(r.instance, "Normal", "UnitTest", fmt.Sprintf("exists is %t", exists))
+			return
+		}
+	}
+
+	// If ism policy is existing do nothing
+	if *r.instance.Status.ExistingISMPolicy {
+		reason = ismPolicyExists
+		return
+	}
+	ismpolicy, retErr := r.CreateISMPolicyRequest()
+	if retErr != nil {
+		reason = "failed to get create the ism policy request"
+		r.logger.Error(retErr, reason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		return
+	}
+	resp, retErr := services.GetPolicy(r.ctx, r.osClient, r.instance.Spec.PolicyID)
+	if retErr != nil {
+		reason = "failed to get policy from Opensearch API"
+		r.logger.Error(retErr, reason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		return
+	}
+	if resp.StatusCode == 404 {
+		r.logger.V(1).Info(fmt.Sprintf("policy %s not found, creating.", r.instance.Spec.PolicyID))
+		retErr = services.CreateISMPolicy(r.ctx, r.osClient, *ismpolicy, r.instance.Spec.PolicyID)
+		if retErr != nil {
+			reason = "failed to create ism policy"
+			r.logger.Error(retErr, reason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			return
+		}
+		r.recorder.Event(r.instance, "Normal", opensearchAPIUpdated, "policy created in opensearch")
+		return ctrl.Result{Requeue: true, RequeueAfter: 30 * time.Second}, retErr
+	}
+	ismResponse, err := services.GetPolicyFromResponse(r.ctx, resp)
+	if err != nil {
+		return
+	}
```

Comment:
This needs to be partially moved into the service. `services.GetPolicy` should directly return a `*requests.Policy` and take care of the logic that is currently in `GetPolicyFromResponse`. The reconciler should never deal with low-level http response types.